xonsh/base_shell.py

```diff
@@ -342,9 +343,14 @@ def precmd(self, line):
         """Called just before execution of line."""
         return line if self.need_more_lines else line.lstrip()
 
-    def default(self, line):
+    def default(self, line, raw_line=None):
         """Implements code execution."""
         line = line if line.endswith("\n") else line + "\n"
+        if not self.need_more_lines:  # this is the first line
+            if not raw_line:
+                self.src_starts_with_space = False
+            else:
+                self.src_starts_with_space = raw_line[0].isspace()
```

Comment:
Two things
1. Does this need to be here?  Can we stick in in the `finally` block before we call to `self._append_history`?
2. It looks like this is set up to only work if the command is a single-line command -- any reason this wouldn't be wanted for a multi-line command?

Reply:
I should probably write a test for the multi line case, but my understanding was that this does work for multiline comands. We only want to set this flag when we're processing the first line, since for subsequent lines, whitespace prefixes have no meaning. If this is the first line, we compute the state of this flag, otherwise we preserve the flag's current state.

Reply:
Ahh, right.  Sorry, I was thinking of the prompt_toolkit backend, where the multiline input is single string, but for readline you'll need to check it the first time.

Reply:
I looked into moving it into the `finally`, but I think that won't work for multi-line commands since the first line would need more input and return early.